src/app/application/services/data-source/DataSource.ts

```diff
@@ -0,0 +1,83 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const get = (src: string, router: Router<Source>) => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
+      }
```

Comment:
**Issue**: This duplicates a polling implementation. Can you please look into whether you can reuse `src/utilities/poll.ts`?

Reply:
I'd rather use generators. That means we can check for whether it should stop in between individual HTTP requests within a single poll loop if we need to (this probably includes backoff/retries etc)

This will replace `src/utilities/poll.ts`, not the other way around.

Reply:
If so, then please externalize it to not prevent other code from using it.

Reply:
Other code could just use an async generator? Apart from that I'm not sure what other use cases we are going to have for polling apart from HTTP requesting?

Reply:
At the moment we only use polling for HTML requests, and as there aren't very many other reasons for polling apart from HTTP requests and we are now going to be using DataSource for all our HTTP requests, as mentioned below

> I’m not a fan of prematurely adding mechanisms that we’re not employing

If for whatever reason we need polling for anything other than HTTP requests, and we can't just use a quick async generator, then we can expose something else. Until then I don't think this is a big deal.

Lastly, I don't want anyone trying to bypass DataSource to do HTTP polling, and giving them an exposed poll function would make it more likely that they try and do that.